addons/web/static/src/js/views/graph/graph_renderer.js

```diff
@@ -1,1107 +1,1031 @@
+    'use strict';
+
+    const config = require('web.config');
+    const core = require('web.core');
+    const dataComparisonUtils = require('web.dataComparisonUtils');
+    const fieldUtils = require('web.field_utils');
+    const OwlAbstractRenderer = require('web.AbstractRendererOwl');
+    const { sortBy } = require('web.utils');
+
+    const _t = core._t;
+    const DateClasses = dataComparisonUtils.DateClasses;
+    const qweb = core.qweb;
+    const { useRef } = owl.hooks;
+
+    const NO_DATA = [_t('No data')];
+    NO_DATA.isNoData = true;
+
+    const COLORS = ["#1f77b4", "#ff7f0e", "#aec7e8", "#ffbb78", "#2ca02c", "#98df8a", "#d62728",
+        "#ff9896", "#9467bd", "#c5b0d5", "#8c564b", "#c49c94", "#e377c2", "#f7b6d2",
+        "#7f7f7f", "#c7c7c7", "#bcbd22", "#dbdb8d", "#17becf", "#9edae5"];
+    const COLOR_NB = COLORS.length;
+
+    function getColor(index) {
+        return COLORS[index % COLOR_NB];
+    }
+
+    function hexToRGBA(hex, opacity) {
+        const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
+        const rgb = result.slice(1, 4).map(n => {
+            return parseInt(n, 16);
+        }).join(',');
+        return 'rgba(' + rgb + ',' + opacity + ')';
+    }
+
+    // used to format values in tooltips and yAxes.
+    const FORMAT_OPTIONS = {
+        // allow to decide if utils.human_number should be used
+        humanReadable: value => {
+            return Math.abs(value) >= 1000;
+        },
+        // with the choices below, 1236 is represented by 1.24k
+        minDigits: 1,
+        decimals: 2,
+        // avoid comma separators for thousands in numbers when human_number is used
+        formatterCallback: str => {
+            return str;
+        },
+    };
+
+    // hide top legend when too many items for device size
+    const MAX_LEGEND_LENGTH = 4 * (Math.max(1, config.device.size_class));
+
+    class GraphRenderer extends OwlAbstractRenderer {
+
+        constructor() {
+            super(...arguments);
+            this.canvasRef = useRef("canvas");
+            this._computeDerivedProps();
         }
-    },
-    /**
-     * This function creates a custom HTML tooltip.
-     *
-     * @private
-     * @param {Object} tooltipModel see chartjs documentation
-     */
-    _customTooltip: function (tooltipModel) {
-        this.$el.css({ cursor: 'default' });
-        if (this.$tooltip) {
-            this.$tooltip.remove();
+
+        mounted() {
+            if (!this.noContentHelperData){
+                this._renderChart();
+            }
         }
-        if (tooltipModel.opacity === 0) {
-            return;
+
+        patched() {
+            if (!this.noContentHelperData){
+                this._renderChart();
+            }
         }
-        if (tooltipModel.dataPoints.length === 0) {
-            return;
+
+        async willUpdateProps(nextProps) {
+            if (this.chart) {
+                this.chart.destroy(); // prevent Chart.js unecessary updates
```

Comment:
Doesn't it produce a flickering? There is a while between the moment willUpdateProps is called and the moment the DOM is patched with the new rendering...

Reply:
I have moved that in patched but I am not comfortable with the idea. Chart.js could react to some action and have incoherent
responses (datasets, data, options are recomputed only in _renderChart)

Reply:
Like if the user clicks on the chart? Of course, if there is an issue because it uses wrong data or something, we have a problem. Calling chart.destroy() removes the chart from the DOM, right? Maybe there is something else we can do to make it inactive?

Reply:
The best solution I have achieved to find has still a flickering when changing mode from line to bar (or conversely). The rest is fine